### PR TITLE
Fmm revision

### PR DESCRIPTION
**Pull request by christianco97**
## Files General Review

> ### 1. FMM fitting files.
#### 1.1 FMM_internal.R
- `replicateGrid` function deleted due to a change in the way to define grids: grids won't change along function calls or depending on backfitting iterations, so only a vector is necessary. Also deleted in `FMM_internal.rd` file.
- `PV` and `PVj` function rewritten (just more compacted code than before).
- `seqTimes` and `angularMean` functions (only a few rewritten lines).
#### 1.2 fitFMM_unit.R
- Checked. No changes.
#### 1.3 fitFMM_back.R
- `replicateGrid` function is not used (see `replicateGrid` above).
#### 1.4 fitFMM.R
- `getApply` is used as supposed. These lines were altered due to restricted FMM fitting problems with parallelization, but it has been fixed.
> ### 2. Other FMM files and functions.
#### 2.1 stopFunctions.R
- in `R2` function, only a few rewritten lines.
#### 2.2 generateFMM.R
- fitted FMM values computation compacted.
- `A`, `alpha`, `beta` and `omega` vector lengths can be not equal: now a **warning** is thrown if those vectors have not the same length.
#### 2.3 plotFMM.R
- Checked. No changes.
#### 2.4 extractWaves.R
- Variable names changed. 
- Use of calculateCosPhi to compute model values.
#### 2.5 getFMMPeaks.R
- Checked. No changes.
> ### 3. FMM class and methods:
- `Class.R` and `FMM-methods.R` checked. No changes.
> ### 4. Other changes.
- Object member access operator (`@`) is used when preparing final FMM object at `fitFMM` function.
- Restricted FMM model fitting has been checked at the previous version (fitFMM_restr.R and FMM_internal_restr).

### CODE REVIEW

File: R/FMM_internal.R [hunk 1]
```diff
@@ -173,62 +170,39 @@ PVj <- function(vData, timePoints, alpha, beta, omega){
```
Comment (alexARC26) [inline R/FMM_internal.R]: Unrelated to the PR: there is a bug in the function `PVj`. It seems to be present even before the PR. See below:

```

library(FMM)
data("ecgData")

fit<-fitFMM(ecgData, nback = 5)
fit; plotFMM(fit)
PVj(ecgData, fit@timePoints, fit@alpha, fit@beta, fit@omega)
PVj(fit@data, fit@timePoints, fit@alpha, fit@beta, fit@omega)
PVj(fit@summarizedData, fit@timePoints, fit@alpha, fit@beta, fit@omega)
# La variabilidad explicada de las componentes 4 y 5 ha cambiado

```

- [x] Why does this happen? Solutions?

File: R/fitFMM_back.R [hunk 2]
```diff
@@ -51,14 +48,14 @@ fitFMM_back<-function(vData, nback, timePoints = seqTimes(length(vData)),
 
       # component j fitting using fitFMM_unit function
       fittedFMMPerComponent[[j]] <- fitFMM_unit(backFittingData, timePoints = timePoints, lengthAlphaGrid = lengthAlphaGrid,
-                                            lengthOmegaGrid = lengthOmegaGrid, alphaGrid = alphaGrid[[j]], omegaMin = omegaMin,
-                                            omegaMax = omegaMax, omegaGrid = omegaGrid[[j]], numReps = numReps, usedApply)
-      fittedValuesPerComponent[,j] <- fittedFMMPerComponent[[j]]@fittedValues
+                                            lengthOmegaGrid = lengthOmegaGrid, alphaGrid = alphaGrid, omegaMin = omegaMin,
```
Comment (alexARC26) [inline R/fitFMM_back.R:51]: Indeed, `alphaGrid` and `omegaGrid` always take the same values if not specified by the user. As this option has been destimated with the latest updates, `replicateGrid` is not needed.

File: R/getFMMPeaks.R [hunk 3]
```diff
@@ -44,31 +44,29 @@ getFMMPeaks <- function(objFMM, timePointsIn2pi = TRUE) {
...
 
   # signal estimation
-  phU <- lapply(1:nComp,function(k) (peakU[k]-alpha)/2)
-  phL <- lapply(1:nComp,function(k) (peakL[k]-alpha)/2)
+  phU <- lapply(1:nComp, function(k) (peakU[k]-alpha)/2)
+  phL <- lapply(1:nComp, function(k) (peakL[k]-alpha)/2)
   ZU <- sapply(1:nComp, function(k) M+sum(A*cos(beta + 2*atan(omega*tan(phU[[k]])))))
   ZL <- sapply(1:nComp, function(k) M+sum(A*cos(beta + 2*atan(omega*tan(phL[[k]])))))
-  names(ZU)<-names(ZL)<-NULL
+  names(ZU) <- names(ZL) <- NULL
```
Comment (alexARC26) [inline R/getFMMPeaks.R]: I find this syntax maybe a bit too fancy 😬
I will change it in the PR review.

**Review by alexARC26 — COMMENTED** [review]
Great! Many changes have been made and the code looks really polished. The changes summary of the PR was really helpful.
Review a bit the changes I have done to see if there is any error. Please, pay attention to the comment I have done about `PV` below and,

- [x] In the revised code, line 34 of `fitFMM_restr` could be the only place where `replicateGrid` was really needed. For now, I have copy-pasted its contents in this function. Look into it, as it may not be neccesary to replicate the grid.

**Comment by christianco97** [comment]
Tomorrow, I will review the package and unless you think otherwise or I find something wrong, I will merge it.

**Comment by alexARC26** [comment]
Okay, I have no errors or bugs to report from my part.

**Pull request merged**